kythe/cxx/indexer/textproto/recordio_textparser.cc

```diff
@@ -0,0 +1,72 @@
+/*
+ * Copyright 2023 The Kythe Authors. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "kythe/cxx/indexer/textproto/recordio_textparser.h"
+
+#include <sstream>
+
+#include "absl/strings/ascii.h"
+#include "absl/strings/match.h"
+#include "absl/strings/str_split.h"
+#include "absl/strings/string_view.h"
+#include "absl/types/optional.h"
+#include "glog/logging.h"
+
+namespace kythe {
+namespace lang_textproto {
+
+void ParseRecordioTextChunks(
+    std::string content, std::string record_delimiter,
+    std::function<void(absl::string_view chunk, int chunk_start_line)>
+        callback) {
+  absl::string_view trimmed_delimiter =
+      absl::StripAsciiWhitespace(record_delimiter);
+  std::vector<std::string> lines = absl::StrSplit(content, absl::ByChar('\n'));
+
+  std::ostringstream chunk;
```

Comment:
Don't copy data to `chunk`; use the StrSplit-emitted string_views directly to calculate the "chunk". As long as you're not copying them around, they will refer to the underlying string_view from the argument and can be used directly.

Any time you would be appending to the stream you should be "extending" the current chunk instead, e.g.

void ExtendBy(absl::string_view& data, std::size_t count) {
  data = absl::string_view(data.data(), data.size() + count);  
}

But you need to ensure that `data` never exceeds the size of the underlying data, this is likely best accomplished by writing your own delimiter for StrSplit which includes the delimiter itself, if found, in the returned line, e.g.
```
struct WithChar {
  explicit WithChar(char ch) : delimiter_(ch) {}
  absl::string_view Find(absl::string_view text, size_t pos) const {
    absl::string_view sep = delimiter_.Find(text, size_t);
    // Always return a zero-width span after the delimiter, so that it's included if present.
    sep.remove_prefix(sep.size()); 
    return sep.
  }
  private:
    absl::ByChar delimiter_;
};
```

Reply:
I am bit confused about the extending "chunk" part. 

Do I need to keep a reference to first line in this case, and keep extending that until I find a record delimiter? Something like this? This is not working as expected for me.
```
absl::string_view chunk;
for (absl::string_view line : absl::StrSplit(content, WithChar('\n'))) {
    if (chunk.empty()) {
      chunk = line;
    }
    if (IsRecordDelimiter(line) {
       callback(...);
       continue;
    }
    if (IsEmptyOrAComment(line)) {
       ExtendBy(chunk, line.size());
       continue;
    }
   ...
}
```

Reply:
Yes, you should be able to transform what you have now into roughly:
```
    if (seen_proto_chunk && (trimmed == trimmed_delimiter)) {
      LOG(INFO) << "Read chunk at " << chunk_begin_line;
      callback(chunk, chunk_begin_line);

      chunk.remove_prefix(chunk.size());
      seen_proto_chunk = false;
      chunk_begin_line = currentline;
      continue;
    }
    if (absl::StartsWith(trimmed, "#") || trimmed.empty()) {
      ExtendBy(chunk, line.size());
      continue;
    }
    // Textproto data, so append it to chunk.
    seen_proto_chunk = true;
    ExtendBy(chunk, line.size());
  }
  // Last line might contains textproto data.
  if (seen_proto_chunk) {
    callback(chunk, chunk_begin_line);
  }
```

Reply:
Thanks! it works now!